src/modm/platform/dma/stm32/dma_hal.hpp.in

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2020, Mike Wolfram
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_STM32_DMA_HAL_HPP
+#define MODM_STM32_DMA_HAL_HPP
+
+#include <modm/architecture/interface/assert.hpp>
+
+#include "dma_base.hpp"
+
+namespace modm
+{
+
+namespace platform
+{
+
+template <uint32_t ID>
+class DmaHal : public DmaBase
+{
+	static_assert(
+%% for instance in dma["instance"]
+    %% if instance == "1"
+    		ID == {{ instance }}
+	%% else
+			or ID == {{ instance }}
+	%% endif
+%% endfor
+			, "invalid DMA controller");
+
+	template <uint32_t id>
+	static constexpr uint32_t getBaseAddress() {
+		if (id == 1)
+			return DMA1_BASE;
+		else
+			return DMA2_BASE;
+	}
+	template <uint32_t id>
+	static constexpr uint32_t getChannelBaseAddress() {
+		if (id == 1)
+			return DMA1_Channel1_BASE;
+		else
+			return DMA2_Channel1_BASE;
+	}
+	template <uint32_t id>
+	static constexpr uint32_t getCselAddress() {
+		if (id == 1)
+			return DMA1_CSELR_BASE;
+		else
+			return DMA2_CSELR_BASE;
+	}
+
+public:
+	static constexpr uint32_t DMA_BASE { getBaseAddress<ID>() };
+	static constexpr uint32_t DMA_CSEL { getCselAddress<ID>() };
+	static constexpr uint32_t CHANNEL_BASE { getChannelBaseAddress<ID>() };
```

Comment:
Why not do the reinterpret cast here and make these of the right pointer type?

Reply:
You cannot initialize a constexpr pointer.

Reply:
Damn \*shakes fist at C++ gods\*